Recursively check for dependent property to fire events.

brisvag opened the pull request:
# Description
`EventedModel` is capable of automatically detecting if a property has a field dependency, and if so, creating an event, and firing it . Consider the following example where `M.a` is used by `M.x` which is in turn used by `M.y`:

```py
from napari.utils.events import EventedModel

class M(EventedModel):
    a = 1

    @property
    def x(self):
        return self.a * 2

    @x.setter
    def x(self, val):
        self.a = val // 2

    @property
    def y(self) -> int:
        return self.x * 10

    @y.setter
    def y(self, val: int):
        self.x = val // 10

m = M()
m.events.connect(lambda event: print(event.type, event.value))

print('> m.a = 2')
m.a = 2
print('> m.x = 3')
m.x = 3
print('> m.y = 50')
m.y = 50
```

If you run it on main, it will print the following:

```
> m.a = 2
a 2
x 4
> m.x = 3
a 1
x 2
x 2
> m.y = 50
a 2
x 4
x 4
y 40
```

As you can see, the `m.y` event is not fired when changing `m.a` or `m.x`; this is because the automatic search for dependent properties is not recursive. This PR changes that, so that the output becomes:

```
> m.a = 2
a 2
x 4
y 40
> m.x = 3
a 1
x 2
y 20
x 2
> m.y = 50
a 2
x 4
y 40
x 4
y 40
```

PS: You probably noticed that we also have an event duplication issue. This can/should be fixed by checking whether an event was already fired, but I'm not sure how to do that yet, and it can be done in a separate PR.


<!-- What does this pull request (PR) do? Why is it necessary? -->
<!-- Tell us about your new feature, improvement, or fix! -->
<!-- If your change includes user interface changes, please add an image, or an animation "An image is worth a thousand words!" -->
<!-- You can use https://www.cockos.com/licecap/ or similar to create animations -->

## Type of change
<!-- Please delete options that are not relevant. -->
- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->
<!-- If this is a bug-fix or otherwise resolves an issue, reference it here with "closes #(issue)" -->

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

brisvag committed:
recursively check for dependents

brisvag committed:
add smart event firing for properties

Czaki commented on napari/utils/events/_tests/test_evented_model.py:
```diff
@@ -461,6 +478,11 @@ def test_evented_model_with_property_setters():
+    assert t.a == 10
 
 
 def test_evented_model_with_property_setters_events():
```
It may be nice to split this test on multiple functions (maybe with mocking in fixture). As it is long and test separate behaviors.

brisvag replied:
I'm not sure how I could convert the mocking into a fixture; could you give me an example?

Czaki replied:
```python
@pytest.fixture()
def mocked_object():
    t = T()
    t.events.a = Mock(t.events.a)
    t.events.b = Mock(t.events.b)
    t.events.c = Mock(t.events.c)
    t.events.d = Mock(t.events.d)
    t.events.e = Mock(t.events.e)
    return t

def test_evented_model_with_property_setters_events(mocked_object):
	assert 'c' in mocked_object.events # the setter has an event

    mocked_object.c = [10, 20]
    mocked_object.events.a.assert_called_with(value=10)
    mocked_object.events.b.assert_called_with(value=20)
    mocked_object.events.c.assert_called_with(value=[10, 20])
    mocked_object.events.d.assert_called_with(value=30)
    mocked_object.events.e.assert_called_with(value=100)

def test_evented_model_with_property_setters_events2(mocked_object):
    mocked_object.d = 8
    mocked_object.events.a.assert_called_with(value=4)
    mocked_object.events.b.assert_called_with(value=4)
    mocked_object.events.c.assert_called_with(value=[4, 4])
    mocked_object.events.d.assert_called_with(value=8)
    mocked_object.events.e.assert_called_with(value=40)
```

etc

brisvag replied:
ok, easier than I thought :P

brisvag replied:
I ended up making it into a parametrized test, I hope it's clear enough.

Czaki commented on napari/utils/events/evented_model.py:
```diff
@@ -108,6 +108,23 @@ def __new__(mcs, name, bases, namespace, **kwargs):
 
+def _update_dependents_from_setter_code(cls, prop, setter, deps, visited=()):
+    """Recursively find all the dependents of a setter by inspecting the code object.
+    Update the given deps dictionary with the new findings.
```
As is understand this not cover code like `setattr(self, "a", 7)`? 
Maybe it is worth to mention this in documentation if this assumption is true.

brisvag replied:
Apparently it does? Wizardry o.o

```py
from napari.utils.events import EventedModel

class M(EventedModel):
    a = 1

    @property
    def x(self):
        return self.a * 2

    @x.setter
    def x(self, val):
        setattr(self, 'a', val // 2)

m = M()
m.events.connect(lambda event: print(event.type, event.value))
print('> m.a = 2')
m.a = 2
```

Czaki replied:
I mean `__field_dependents__` calculation, not triggering events.

brisvag replied:
I'm not sure I follow... Triggering events is based on the `__field_dependets__` content, so if that works it means the former worked.

Czaki replied:
Ok. Here Is something what I have in mind:

```python
from napari.utils.events import EventedModel

class M(EventedModel):
    a = 1

    @property
    def x(self):
        return getattr(self, "a") * 2

    @x.setter
    def x(self, val):
        setattr(self, 'a', val // 2)

    @property
    def y(self):
        return self.a * 3

    @y.setter
    def y(self, val):
        self.a = val // 3


print(M.__field_dependents__)
```
The output is 

```
{'a': {'y'}}
```

brisvag replied:
Ah sorry, I only changed the setter, my bad :P Yeah this is a caveat, but I don't think there's a sensible way around it. In that case, you can just declare dependencies explicitly with `Config`:

```py
from napari.utils.events import EventedModel

class M(EventedModel):
    a = 1

    @property
    def x(self):
        return getattr(self, "a") * 2

    @x.setter
    def x(self, val):
        setattr(self, 'a', val // 2)

    @property
    def y(self):
        return self.a * 3

    @y.setter
    def y(self, val):
        self.a = val // 3

    class Config:
        dependencies={
            'x': ['a'],
            'y': ['a']
        }


print(M.__field_dependents__)
```

brisvag replied:
Added some docstring.

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into fix/evented-dependent-properties

brisvag committed:
parametrize messy event

Czaki commented on napari/utils/events/_tests/test_evented_model.py:
```diff
@@ -461,31 +478,43 @@ def test_evented_model_with_property_setters():
+)
+def test_evented_model_with_property_setters_events_1(
+    mocked_object, attribute, value, expected_event_values
+):
```
On the one side, it is more compacted, from the other side, it is less readable. 

Adding a docstring with a test explanation for the current code will be nice.

brisvag committed:
better names and added caveat

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

brisvag committed:
add test docstring

brisvag committed:
Merge branch 'fix/evented-dependent-properties' of github.com:brisvag/napari into fix/evented-dependent-properties

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into fix/evented-dependent-properties

brisvag committed:
Merge branch 'fix/evented-dependent-properties' of github.com:brisvag/napari into fix/evented-dependent-properties

brisvag committed:
Merge branch 'main' into fix/evented-dependent-properties

brisvag committed:
Merge branch 'main' into fix/evented-dependent-properties

Czaki commented on napari/utils/events/evented_model.py:
```diff
@@ -139,23 +175,21 @@ class Config:
                 if field not in cls.__fields__:
                     warnings.warn(f"Unrecognized field dependency: {field}")
-                deps.setdefault(field, set()).add(prop)
+                deps.setdefault(field, set()).add(prop_name)
```
Maybe some test to cover this? 
EventedModel is the core of napari so I would like to see 100% coverage

Czaki commented on napari/utils/events/evented_model.py:
```diff
@@ -234,19 +268,33 @@ def __setattr__(self, name: str, value: Any) -> None:
         # grab current value
         before = getattr(self, name, object())
+        before_deps = {}
+        with warnings.catch_warnings():
+            # we still need to check for deprecated properties
+            warnings.simplefilter("ignore", DeprecationWarning)
```
As catch_warnings is not thread-safe, there is a need to work around it. But as it will not be simple, then there is a need to open an Issue describing the problem and solve it in a follow-up PR. 

My idea is to define `deprecated_property` that will use instance-specific variable to control if emit deprecation/future warning.

brisvag committed:
make it work without setters and add tests

Carreau committed:
Remove manual binding of values.

No need to manually add the values event as now every property is
already tracked.

brisvag committed:
Merge branch 'main' into fix/evented-dependent-properties

brisvag committed:
fix property comparison and test

Czaki commented on napari/utils/events/_tests/test_evented_model.py:
```diff
@@ -461,31 +475,101 @@ def test_evented_model_with_property_setters():
+            a: int = 1
+
+            @property
+            def b(self):
```
```suggestion
            def b(self):  # pragma: no cover
```

Czaki commented on napari/utils/events/_tests/test_evented_model.py:
```diff
@@ -461,31 +475,101 @@ def test_evented_model_with_property_setters():
+            a: int = 1
+
+            @property
+            def b(self):
```
```suggestion
            def b(self):  # pragma: no cover
```

Czaki reviewed (approved):
Looks good now.

brisvag committed:
Apply suggestions from code review

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into fix/evented-dependent-properties
